### DISCUSSION TITLE
pymeasure.instruments.keithley.Keithley2400.means() not returning expected data

### DISCUSSION THREAD

**Discussion by rob-robinson-14** [pymeasure/pymeasure#681] [Q&A]
I have a script that is performing lots of measurements. The majority of the time the script behaves as expected however I get instances where self.means() returns an array that is larger or smaller than 3 and cannot work out why and I am hoping someone may be able to help me out. 

My measurement code:
`

    def measure (self):
      
        self.disable_buffer()
        self.reset_buffer()
        self.start_buffer()
        self.trigger_immediately()
        try:
            while len(self.buffer_data) != self.buffer_points:
                time.sleep(0.001)
            results = self.means
            if len(results) != 3:
                print(f"results: {results}")
                print(f"means: {self.means}")
                print("Length Error")
                return [0, 0, 0]
            return results # [voltage,current,resistance]
        except:
            print("Keithely Measure Error")
            return [0, 0, 0]

`

When the data length != 3 here are some of the outputs:

```
results: [0.1482621, 0.1456199, 0.1430148, 0.1417649, 0.1376793, 0.1346239, 0.1334049, 0.1325033, 0.1292747, 0.1249919]
means: 10.0
Length Error

```

and 

```
results: [0.1777953, 0.1854779, 0.2163265, 0.2148014, 0.2129671, 0.2153001, 0.213824, 0.214906, 0.2137784, '+2.141559E-01;10']
means: [0.2079333, 0.0001953703, 9.91e+37]
Length Error
```

The erroneous returned occurred twice in 300 measurements. 

It seems that the self.means is first returning the buffer and then the buffer_points. Perhaps I need a delay after the buffer_data is collected and before I call self.means() but any change I make would be guess work so wondering if anyone can offer some advice with a bit more justification behind it. 

FYI - This is my setup for the SMU

```
    def smu_setup(self):
        
        # Compliance voltage must be greater than open source voltage for PV cell
        self.smu.apply_current(current_range=0.1,compliance_voltage=7)
        self.smu.source_current = self.IphLevels[0]
        self.smu.compliance_voltage = 7
        self.smu.use_front_terminals()
        self.smu.measure_current(nplc=0.1, current=0.2e-3, auto_range=True)
        self.smu.measure_voltage(nplc=0.1, voltage=7, auto_range=True)
        self.smu.measure_concurent_functions = True

        # self.smu.filter_count = 10
        # self.smu.filter_type = 'REP' # REP : Repeating filter MOV : Moving filter
        # self.smu.filter_state = 'ON'



        self.smu.disable_buffer() #Disables the connection between measurements and the buffer, but does not abort the measurement process.
        self.smu.stop_buffer()
        self.smu.buffer_points = 10
        self.smu.filter_count = 5
        self.smu.filter_type = 'REP' # REP : Repeating filter MOV : Moving filter
        self.smu.filter_state = 'ON'
        self.smu.config_buffer(self.smu.buffer_points)
        
        self.smu.enable_source()
        return

```

Cheers,
Rob

**Comment by bilderbuchi** [comment]
Could be your instrument gets saturated by too quick requests for `self.buffer_data`, and somehow one too many ends up in the read buffer. I'd first try increasing that sleep duration and see if that fixes the problem. 
Slightly related, doesn't the instrument offer a way to signal when the buffer is full (wait for buffer or somesuch)? Maybe if you can wrap that in a short method that would be better than doing this "by hand"

**Reply by rob-robinson-14** [reply]
Thank you for your response.

Great thanks I will give that a go. 

Yes there is a method "is_buffer_full()" but I was finding that it never returned True even after the measurements had finished and so used the above as a work around.
